lib/jxl/render_pipeline/write_to_ib_stage.cc

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include "lib/jxl/render_pipeline/write_to_ib_stage.h"
+
+#include "lib/jxl/common.h"
+#include "lib/jxl/image_bundle.h"
+
+namespace jxl {
+
+namespace {
+class WriteToImageBundleStage : public RenderPipelineStage {
+ public:
+  explicit WriteToImageBundleStage(ImageBundle* image_bundle)
+      : RenderPipelineStage(RenderPipelineStage::Settings()),
+        image_bundle_(image_bundle) {}
+
+  void ProcessRow(const RowInfo& input_rows, const RowInfo& output_rows,
+                  size_t xextra, size_t xsize, size_t xpos, size_t ypos,
+                  float* JXL_RESTRICT temp) const final {
+    for (size_t c = 0; c < 3; c++) {
+      memcpy(image_bundle_->color()->PlaneRow(c, ypos) + xpos - xextra,
+             GetInputRow(input_rows, c, 0) + kRenderPipelineXOffset - xextra,
+             sizeof(float) * (xsize + 2 * xextra));
+    }
+    for (size_t ec = 0; ec < image_bundle_->extra_channels().size(); ec++) {
+      memcpy(
+          image_bundle_->extra_channels()[ec].Row(ypos) + xpos - xextra,
+          GetInputRow(input_rows, 3 + ec, 0) + kRenderPipelineXOffset - xextra,
+          sizeof(float) * (xsize + 2 * xextra));
+    }
+  }
+
+  RenderPipelineChannelMode GetChannelMode(size_t c) const final {
+    return c < 3 + image_bundle_->extra_channels().size()
+               ? RenderPipelineChannelMode::kInput
+               : RenderPipelineChannelMode::kIgnored;
```

Comment:
Is that really needed? Non-existing channels will always be ignored by any stage, can't we just have an invariant that GetChannelMode() does not get called with an out-of-bounds channel index?

Or is this preparing for the scenario where the output image bundle has fewer extra channels than the actual jxl image, since some extra channels are "internal only"? (or not needed to render the image, e.g. a kDepth channel)

In that case, it makes sense to keep it like this, but add a comment that the output image bundle can have fewer extra channels than the input, semantics being that only the first N are written.

Reply:
At some point, we'll have noise patterns as pipeline "channels" :)